hugr-passes/src/dataflow.rs

```diff
@@ -0,0 +1,91 @@
+#![warn(missing_docs)]
+//! Dataflow analysis of Hugrs.
+
+mod datalog;
+
+mod machine;
+use hugr_core::ops::constant::CustomConst;
+pub use machine::{Machine, TailLoopTermination};
+
+mod partial_value;
+pub use partial_value::{AbstractValue, PartialSum, PartialValue, Sum};
+
+use hugr_core::ops::{ExtensionOp, Value};
+use hugr_core::{Hugr, Node};
+use std::hash::Hash;
+
+/// Clients of the dataflow framework (particular analyses, such as constant folding)
+/// must implement this trait (including providing an appropriate domain type `V`).
+pub trait DFContext<V>: Clone + Eq + Hash + std::ops::Deref<Target = Hugr> {
+    /// Given lattice values for each input, update lattice values for the (dataflow) outputs.
+    /// For extension ops only, excluding [MakeTuple] and [UnpackTuple].
+    /// `_outs` is an array with one element per dataflow output, each initialized to [PartialValue::Top]
+    /// which is the correct value to leave if nothing can be deduced about that output.
+    /// (The default does nothing, i.e. leaves `Top` for all outputs.)
+    ///
+    /// [MakeTuple]: hugr_core::extension::prelude::MakeTuple
+    /// [UnpackTuple]: hugr_core::extension::prelude::UnpackTuple
+    fn interpret_leaf_op(
+        &self,
+        _node: Node,
+        _e: &ExtensionOp,
+        _ins: &[PartialValue<V>],
+        _outs: &mut [PartialValue<V>],
+    ) {
+    }
+
+    /// Produces an abstract value from a constant. The default impl
+    /// traverses the constant [Value] to its leaves ([Value::Extension] and [Value::Function]),
+    /// converts these using [Self::value_from_custom_const] and [Self::value_from_const_hugr],
+    /// and builds nested [PartialValue::new_variant] to represent the structure.
+    fn value_from_const(&self, n: Node, cst: &Value) -> PartialValue<V> {
```

Comment:
Alternatives include (a) moving `value_from_const` out of the trait and fixing it; I don't have a concrete reason why anyone would need to override it but it seems "reasonable" that one might; or (b) moving `value_from_custom_const` and `value_from_const_hugr` into a separate trait, and then implementing `value_from_const` for any type providing those two - this would be the more proper way of doing what I've done here but feels a bit overkill TBH.